### PR TITLE
Fix abort in glpk test

### PR DESCRIPTION
**Pull request by cxzhong**
## Problem

Reentrant GLPK builds keep allocator environment state in native compiler TLS (`_Thread_local` or `__thread`). A `glp_prob` or `glp_graph` allocated on one native thread must therefore be deleted while that thread's GLPK environment is active.

Python may run a Cython extension object's deallocator on a different thread. Deleting the GLPK pointer there resolves a different allocator environment, corrupts its allocation bookkeeping, and can abort with `glp_free: memory allocation error`.

Fixes #28559.

## Root cause

The old backends treated a GLPK pointer as independent of the native thread that created it. Python thread identifiers alone are not sufficient for ownership because native IDs can be recycled, and GLPK does not install a destructor that releases outstanding objects when its TLS environment exits.

The same ownership rule applies to both LP and graph objects. Cleanup must happen before `glp_free_env()` on the owner thread, while attempts to use a pointer from any other thread must fail without calling into GLPK.

## Changes

- Add a shared per-thread `GLPKThreadResources` registry for LP and graph pointers, backed by a private thread-local cleanup guard.
- Record the native thread ID together with Python interpreter and thread-state identities so recycled native IDs cannot impersonate an owner.
- Delete same-thread objects immediately, defer foreign-thread destruction to the live owner, and sweep deferred objects on the owner's next LP or graph operation.
- On normal owner-thread exit, delete registered GLPK objects, clear their wrapper pointers, and only then release the GLPK environment.
- Detect process-global GLPK builds with `glp_config("TLS")`; serialized cross-thread handoff remains supported when no thread-local environment is used.
- Make registration failure-safe so a failed registry append cannot leave a cleanup guard that later frees the same pointer twice.
- Validate pointer ownership before every GLPK access and raise `RuntimeError` for unavailable or foreign-thread backends.
- Make the generic boolean backend query contract exception-capable with `except? -1`, and update the in-tree implementations without breaking separately distributed COIN, Gurobi, or CPLEX backend sources.
- Reuse validated pointers in bulk/hot paths and add regression coverage for negative graph optima and invalid row-dual indices.

## Resulting behavior

| Situation | Behavior |
| --- | --- |
| Access on the creating thread | Normal GLPK operation |
| Access from another thread | Raises `RuntimeError` before entering GLPK |
| Last Python reference is dropped elsewhere while the owner remains alive | Destruction is deferred and swept by the owner's next LP/graph operation or thread exit |
| Owner thread exits while a wrapper survives | The pointer is deleted and cleared before the owner's GLPK environment is released; later access raises `RuntimeError` |
| GLPK uses a process-global environment | Serialized cross-thread handoff remains supported |

## Validation

- Rebuilt all 33 affected Cython targets with `ninja -C build/cp312`.
- `glpk_backend_test.py`: 12 passed, 1 skipped (the process-global-only case on this TLS build).
- Focused doctests: 1,720 passed across the GLPK resource/backend/graph modules, MIP, and the generic backend.
- Broader affected-backend doctests: 942 active tests passed.
- Sage coverage: 100% for `glpk_backend.pyx`, `glpk_graph_backend.pyx`, and `glpk_thread_resources.py`.
- Cythonized the current external COIN, Gurobi, and CPLEX backend sources against the revised declarations.
- `ruff check` and `git diff --check` pass.

## Dependencies

None.

## Checklist

- [x] Added regression tests for the bug and lifecycle edge cases.
- [x] Added documentation and doctests for the new resource-management API.
- [x] Linked the relevant issue.

### CODE REVIEW

**Comment by orlitzky** [comment]
This was probably triggered by putting `-ansi` in the `CFLAGS` for GLPK. Try `-std=gnu17`.

**Comment by orlitzky** [comment]
If that doesn't work, `--enable-reentrant` or `--disable-reentrant`, whichever one is not what gets picked by ./configure, may help.

**Comment by cxzhong** [comment]
This also happens before https://github.com/sagemath/sage/issues/28559
I think something tirggers this again

**Comment by orlitzky** [comment]
In that case, if the analysis is right, `--disable-reentrant` should prevent thread-local storage from being used. We don't need a good solution, we just need to keep the CI passing until we can deprecate and remove GLPK.

**Comment by cxzhong** [comment]
But we use the system's glpk now.
```
#24 23.17 glpk:                           using system package; SPKG will not be installed
```
So I think maybe the rebuild with gcc-15 let the hidden problem happens again
```
diff -Nru glpk-5.0/debian/changelog glpk-5.0/debian/changelog
--- glpk-5.0/debian/changelog	2025-08-11 13:52:03.000000000 +0000
+++ glpk-5.0/debian/changelog	2026-02-03 22:10:42.000000000 +0000
@@ -1,3 +1,9 @@
+glpk (5.0-2build1) resolute; urgency=medium
+
+  * No-change mass rebuild for Ubuntu 26.04 (LP: #2132257)
+
+ -- Sebastien Bacher <seb128@debian.org>  Tue, 03 Feb 2026 23:10:42 +0100
+
 glpk (5.0-2) unstable; urgency=medium
 
   * gcc-15.patch: new patch, fixes FTBFS against GCC 15 (Closes: #1096728)
```

**Comment by cxzhong** [comment]
And in conda's build, we still do not use this parameter.

**Comment by orlitzky** [comment]
Patching for `bool` was probably not the way to go. Using `-ansi` prevents distros from unbundling suitesparse, but `-std=gnu17` is OK: https://github.com/gentoo/gentoo/blob/master/sci-mathematics/glpk/glpk-5.0-r4.ebuild#L36

I wish I remembered what the comment about `-std=c17` was about.

I'm not trying to be difficult, but this is a very delicate issue involving memory corruption across multiple threads that depends on build-time options and CFLAGS in an external dependency. It will take a long time for someone to review properly... and the GLPK backend isn't even used now!

Maybe we leave this as a "CI fix" and deprecate the GLPK backend immediately? It will still prevent the CI from failing even if we don't merge it. (I don't personally care if we trash the GLPK backend right now, but do not want to put my name next to a positive review that I haven't bothered to look at.)

**Comment by cxzhong** [comment]
I think, we may change `glpk` to optional next time. `glpk_graph_backend` I think we can not replace by highs?

I hope this can be merged

**Comment by orlitzky** [comment]
> I think, we may change `glpk` to optional next time. `glpk_graph_backend` I think we can not replace by highs?

There is nothing to replace, it is unused: https://github.com/sagemath/sage/pull/42271

**Comment by cxzhong** [comment]
So we merge this and deprecate the glpk module now? Then we do not need to care this part

**Comment by cxzhong** [comment]
I do not think someone use glpk instead of highs

**Comment by orlitzky** [comment]
> So we merge this and deprecate the glpk module now? Then we do not need to care this part

For me it is a matter of procedure, I don't want to say "I have reviewed this and it looks OK to me" if I don't understand what the changes do. (It would take time that I don't have right now to understand them.)

What if we deprecate these two modules, and skip them in the CI from now on? We are already using `sage -t --exclude` to skip some flaky files, it should be possible to ignore the GLPK modules as well. That I could review in a few minutes.

**Comment by dimpase** [comment]
there is some nontrivially looking multithreaded Cython code. Doesn't look very easy.
